platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,148 @@
+# Content Validation
+
+**Author(s):** Bill Fienberg
+**Last Updated:** May 26 2020  
+**Status:** **Draft** | In Review | Approved
+**Approvers:** Dror Matalon [], Andrew Gunsch [], Rian Fowler [], Steve Wirt []
+
+## Table of Contents
+
+- [Content Validation](#content-validation)
+  - [Table of Contents](#table-of-contents)
+  - [Overview](#overview)
+    - [Objective](#objective)
+    - [Background](#background)
+      - [Broken Link Checking](#broken-link-checking)
+      - [Accessibility Checking](#accessibility-checking)
+      - [Pain points](#pain-points)
+        - [Content Writers](#content-writers)
+        - [Users](#users)
+        - [FE Tools team](#fe-tools-team)
+    - [High Level Design](#high-level-design)
+  - [Specifics](#specifics)
+    - [Detailed Design](#detailed-design)
+    - [Code Location](#code-location)
+    - [Testing Plan](#testing-plan)
+    - [Logging](#logging)
+    - [Debugging](#debugging)
+    - [Caveats](#caveats)
+    - [Security Concerns](#security-concerns)
+    - [Privacy Concerns](#privacy-concerns)
+    - [Open Questions and Risks](#open-questions-and-risks)
+    - [Work Estimates](#work-estimates)
+    - [Alternatives](#alternatives)
+      - [Highest quality, but lowest speed](#highest-quality-but-lowest-speed)
+      - [Fastest speed, but lowest quality](#fastest-speed-but-lowest-quality)
+    - [Future Work](#future-work)
+    - [Revision History](#revision-history)
+
+## Overview
+
+### Objective
+
+The objective here is to define an approach to validate the CMS content. Specifically, this design doc focuses on validating that CMS content has no accessibility errors and/or broken links.
+
+This design document is intended for front end and DevOps engineers on the Veteran-facing Services Platform (VSP) and Veteran-Facing Services (VFS) teams.
+
+### Background
+
+The `vets-website` repo currently contains [one script that builds both the content and the applications](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/index.js). Chris V. wrote a [design doc to disentangle the content build from the application build](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md), and that work is currently in progress.
+
+Right now, content validation consists of broken link checking and accessibility checking.
+
+##### Broken Link Checking
+
+The `check-broken-link` plugin does the following:
+
+1. Loop through all of the HTML files in the Metalsmith pipeline
+2. Extract all of the broken `href`/`src` values using the `getBrokenLinks` helper
+3. Formats the result into useful console output
+4. Breaks the build on production, and logs the output on lower environments
+
+##### Accessibility Checking
+
+
```

Comment:
When the integration step of the build fails, there is a generic notification sent to the #devops-deploys channel via https://github.com/department-of-veterans-affairs/vets-website/blob/f29aa3a385d03479308f78c0c22c292ecae93f20/Jenkinsfile#L86, correct? If so, the violation must be reported manually because the error message isn't descriptive, correct?

For example: https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800

Reply:
I believe that's all correct, yeah.